File: blind75/44_spiral_matrix/spiral_matrix.py
from typing import List


def spiral_order(matrix: List[List[int]]) -> List[int]:
    """
    Returns all elements of the matrix in spiral order.
    Uses boundary tracking for O(1) space complexity.

    Args:
        matrix: 2D list of integers

    Returns:
        List of integers in spiral order
    """
    if not matrix or not matrix[0]:
        return []

    m, n = len(matrix), len(matrix[0])
    result = []
    top, bottom, left, right = 0, m - 1, 0, n - 1

    while top <= bottom and left <= right:
        # Move right along top row
        for j in range(left, right + 1):
            result.append(matrix[top][j])
        top += 1

        # Move down along right column
        for i in range(top, bottom + 1):
            result.append(matrix[i][right])
        right -= 1

        # Move left along bottom row (if there's a row remaining)
        if top <= bottom:
            for j in range(right, left - 1, -1):
                result.append(matrix[bottom][j])
            bottom -= 1

        # Move up along left column (if there's a column remaining)
        if left <= right:
            for i in range(bottom, top - 1, -1):
                result.append(matrix[i][left])
            left += 1

    return result

File: blind75/44_spiral_matrix/spiral_matrix.go
package spiral_matrix

// SpiralOrder returns all elements of the matrix in spiral order.
// Uses boundary tracking for O(1) space complexity.
func SpiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return []int{}
	}

	m, n := len(matrix), len(matrix[0])
	result := make([]int, 0, m*n)
	top, bottom, left, right := 0, m-1, 0, n-1

	for top <= bottom && left <= right {
		// Move right along top row
		for j := left; j <= right; j++ {
			result = append(result, matrix[top][j])
		}
		top++

		// Move down along right column
		for i := top; i <= bottom; i++ {
			result = append(result, matrix[i][right])
		}
		right--

		// Move left along bottom row (if there's a row remaining)
		if top <= bottom {
			for j := right; j >= left; j-- {
				result = append(result, matrix[bottom][j])
			}
			bottom--
		}

		// Move up along left column (if there's a column remaining)
		if left <= right {
			for i := bottom; i >= top; i-- {
				result = append(result, matrix[i][left])
			}
			left++
		}
	}

	return result
}
